controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1199 @@
+		log.Info("using explicit host", "host.id", hostID)
+		return s.Host(hostID)
+	}
+
+	counts := s.jobs.GetHostJobCounts(formation.key(), typ)
+	var minCount int = math.MaxInt32
+	for _, host := range hosts {
+		count, ok := counts[host.ID()]
+		if !ok || count < minCount {
+			minCount = count
+			hostID = host.ID()
+		}
+	}
+	log.Info(fmt.Sprintf("using host with least %s jobs", typ), "host.id", hostID)
+	return s.Host(hostID)
+}
+
+func (s *Scheduler) getHosts() ([]utils.HostClient, error) {
+	hosts, err := s.Hosts()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that we're only following hosts that we can discover
+	knownHosts := make(map[string]struct{})
+	for id, hostStream := range s.hostStreams {
+		if hostStream.Err() == nil {
+			knownHosts[id] = struct{}{}
+		} else {
+			s.unfollowHost(id)
+		}
+	}
+	for _, h := range hosts {
+		delete(knownHosts, h.ID())
+		s.followHost(h)
+	}
+	for id := range knownHosts {
+		s.unfollowHost(id)
+	}
+	return hosts, nil
+}
+
+func (s *Scheduler) Stop() error {
+	log := fnLogger()
+	log.Info("stopping scheduler loop")
+	s.stopOnce.Do(func() {
+		close(s.stop)
+	})
+	return nil
+}
+
+func (s *Scheduler) Subscribe(events chan Event) stream.Stream {
+	log := fnLogger()
+	log.Info("adding event subscriber")
+	s.listenMtx.Lock()
+	defer s.listenMtx.Unlock()
+	s.listeners[events] = struct{}{}
+	return &Stream{s, events}
+}
+
+func (s *Scheduler) Unsubscribe(events chan Event) {
+	log := fnLogger()
+	log.Info("removing event subscriber")
+	s.listenMtx.Lock()
+	defer s.listenMtx.Unlock()
+	delete(s.listeners, events)
+}
+
+func (s *Scheduler) SaveJob(job *Job, appName string, status host.JobStatus, metadata map[string]string) {
+	log := fnLogger("job.id", job.JobID, "app.id", job.AppID, "app.name", appName, "release.id", job.ReleaseID, "job.type", job.Type, "job.status", status)
+	log.Info("queuing job for persistence")
+	s.putJobs <- controllerJobFromSchedulerJob(job, jobState(status), metadata)
+	if job.Formation == nil {
+		return
+	}
+	switch status {
+	case host.StatusStarting:
+		fallthrough
+	case host.StatusRunning:
+		s.handleJobStart(job)
+	default:
+		if job.IsScheduled() {
+			s.handleJobCrash(job)
+		} else {
+			s.handleJobStop(job)
+		}
+	}
+}
+
+func (s *Scheduler) Jobs() map[string]*Job {
+	jobs := make(map[string]*Job)
+	for id, j := range s.jobs {
+		if j.state == JobStateUp {
+			jobs[id] = j
+		}
+	}
+	return jobs
+}
+
+func (s *Scheduler) scheduleJobStart(job *Job) error {
+	log := fnLogger()
+	if !s.isLeader {
+		return errors.New("this scheduler is not the leader")
+	}
+	if job.startedAt.Before(time.Now().Add(-s.backoffPeriod)) {
+		log.Info("resetting job restarts", "backoffPeriod", s.backoffPeriod, "job.startedAt", job.startedAt)
+		job.restarts = 0
+	}
+	backoff := s.getBackoffDuration(job.restarts)
+	if backoff == s.maxBackoffPeriod {
+		log.Warn("reached maximum backoff period, allowing job to stay crashed")
+		return errors.New("maximum backoff period reached")
+	}
```

Comment:
Ah ok, in that case why would we want to start it again straight away (via rectify) rather than continuing at the maximum backoff? Lets stick closely to the [Heroku policy](https://devcenter.heroku.com/articles/dynos#dyno-crash-restart-policy), which is what our policy is based on.